proxy/src/main/java/com/wavefront/agent/handlers/TrafficShapingRateLimitAdjuster.java

```diff
@@ -0,0 +1,73 @@
+package com.wavefront.agent.handlers;
+
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.Timer;
+import java.util.TimerTask;
+import java.util.concurrent.atomic.AtomicLong;
+
+import com.google.common.util.concurrent.RecyclableRateLimiter;
+import com.wavefront.agent.data.EntityPropertiesFactory;
+import com.wavefront.common.Managed;
+import com.wavefront.data.ReportableEntityType;
+import com.yammer.metrics.stats.ExponentiallyDecayingSample;
+import com.yammer.metrics.stats.Sample;
+
+/**
+ * Experimental: use automatic traffic shaping (set rate limiter based on last 5 minutes of
+ * received rates
+ *
+ * @author vasily@wavefront.com.
+ */
+public class TrafficShapingRateLimitAdjuster extends TimerTask implements Managed {
+  private static final int DEFAULT_SAMPLE_SIZE = 1028;
```

Comment:
any reason for 1028 and not 1024?

Reply:
1028 samples with 0.015 alpha seem to be the industry standard, which heavily biases the reservoir to the last 5 minutes. 

> also what happens when the rate limit hits? does it queue or reject the point on the listener?

Eventually we queue, but only after we exhaust in-memory buffers. I tried the "auto adjust the flush batch size" approach, that doesn't use the rate limiter, at first; but that didn't give us enough flexibility in managing traffic. We would have to dramatically increase the size of in-memory buffers, and even then, if(when) the buffer limit is breached, the entire contents of the buffer would be immediately drained to the queue. Which would in turn immediately trigger backlog processing threads so in the end it would basically cause an even larger burst of traffic, which would defeat the purpose.